app/src/main/java/bg/zahov/app/ui/workout/WorkoutViewModelNew.kt

```diff
@@ -0,0 +1,905 @@
+     *
+     * This function creates a new set entry. If a corresponding template exercise exists and contains predefined sets,
+     * it uses the template data to populate the set. Otherwise, it creates a default set entry.
+     */
+    private fun insertSetAtIndex(
+        exercises: MutableList<WorkoutEntry>,
+        insertIndex: Int,
+        exercisePosition: Int,
+        exerciseCategory: Category,
+        templateExercise: Exercise?,
+    ) {
+        val setNumber = insertIndex - exercisePosition
+
+        val setEntry = if (templateExercise != null && setNumber < templateExercise.sets.size) {
+            templateExercise.sets[setNumber].toSetEntry(
+                exerciseCategory = templateExercise.category,
+                setNumber = setNumber - 1,
+                previousResults = "${templateExercise.sets[setNumber].secondMetric} x ${templateExercise.sets[setNumber].secondMetric}"
+            )
+        } else {
+            WorkoutEntry.SetEntry(
+                firstInputFieldVisibility = exerciseCategory != Category.None,
+                setNumber = setNumber.toString(),
+                set = Sets(type = SetType.DEFAULT, firstMetric = 0.0, secondMetric = 0),
+                previousResults = "-/-"
+            )
+        }
+
+        exercises.add(insertIndex, setEntry)
+    }
+
+    /**
+     * Changes the note for the workout
+     *
+     * @param note The new note for the workout
+     */
+    fun changeNote(note: String) {
+        _uiState.update { old ->
+            old.copy(note = note)
+        }
+    }
+
+    /**
+     * Removes the set in the specified position
+     * and renumbers sets above/below it if needed
+     *
+     * @param position The index of the set to remove
+     */
+    fun removeSet(position: Int) {
+        _uiState.update { old ->
+            val exercises = old.exercises.toMutableList()
+            exercises.removeAt(position)
+            var index = position
+            while (index < exercises.size && exercises[index] is WorkoutEntry.SetEntry) {
+                (exercises[index] as? WorkoutEntry.SetEntry)?.let { setEntry ->
+                    exercises[index] = setEntry.copy(
+                        setNumber = (setEntry.setNumber.toInt() - 1).toString()
+                    )
+                }
+                index++
+            }
+            old.copy(exercises = exercises)
+        }
+    }
+
+    /**
+     * based on which field has changed
+     * writes the value in the appropriate field
+     *
+     * @param position The index of the set for whom we want to change the value
+     * @param metric The value we want to update
+     * @param setField We use it to determine whether we want to write to
+     * firstMetric - [SetField.REPETITIONS] or secondMetric - [SetField.WEIGHT].
+     */
+    fun onInputFieldChanged(
+        position: Int,
+        metric: String,
+        setField: SetField,
```

Comment:
why not create separate functions for adding weigh and repetitions, I know you want to avoid code duplication, but it will be more clear to have separate ones

Reply:
🫡